arcs/src/systems/spatial_relation.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{
+    components::{SpatialEntity, DrawingObject, Space},
+    algorithms::{Bounded},
+};
+use specs::prelude::*;
+
+/// A [`System`] which keeps track of the spatial relation of entities
+#[derive(Debug)]
+pub struct SpatialRelation {
+    changes: ReaderId<ComponentEvent>,
+    to_insert: BitSet,
+    to_update: BitSet,
+}
+
+impl SpatialRelation {
+    pub const NAME: &'static str = module_path!();
+
+    pub fn new(world: &World) -> Self {        
+        SpatialRelation {
+            changes: world.write_storage::<DrawingObject>().register_reader(),
+            to_insert: BitSet::new(),
+            to_update: BitSet::new(),
+        }
+    }
+}
+
+impl<'world> System<'world> for SpatialRelation {
+    type SystemData = (
+        Write<'world, Space>,
+        ReadStorage<'world, DrawingObject>,
+        Entities<'world>
+    );
+
+    fn run(&mut self, data: Self::SystemData) {
+        // clear any left-over flags
+        self.to_insert.clear();
+        self.to_update.clear();
+
+        let (mut space, drawing_objects, entities) = data;
+
+        // find out which items have changed since we were last polled
+        for event in drawing_objects.channel().read(&mut self.changes) {
+            println!("Event is {:#?}", event);
+            match *event {
+                ComponentEvent::Inserted(id) => {
+                    self.to_insert.add(id);
+                },
+                ComponentEvent::Modified(id) => {
+                    self.to_update.add(id);
+                },
+                ComponentEvent::Removed(id) => {
+                    space.remove_by_id(id);
+                },
+            }
+        }
+
+        for (ent, drawing_object, _) in
+            (&entities, &drawing_objects, &self.to_insert).join()
+        {
+            let bb = drawing_object.geometry.bounding_box();
+            space.insert(SpatialEntity::new(bb, ent));
+        }
+
+        for (ent, drawing_object, _) in
+            (&entities, &drawing_objects, &self.to_update).join()
+        {
+            let bb = drawing_object.geometry.bounding_box();
+            space.modify(SpatialEntity::new(bb, ent));
+        }
+    }
+
+    fn setup(&mut self, world: &mut World) {
+        <Self::SystemData as shred::DynamicSystemData>::setup(
+            &self.accessor(),
+            world,
+        );
+
+        let drawing_storage = world.read_storage::<DrawingObject>();
+        let mut space = world.write_resource::<Space>();
+
+        space.clear();
+
+        for entity in world.entities().join() {
```

Comment:
I don't think we want to iterate over *every* `Entity` in the world, you can iterate over just the items with a` DrawingObject` component using `for (entity, drawing_object) in (world.entities(), &drawing_storage).join() { ... }`.

Reply:
Done